cli/lib/nf3.mjs

```diff
@@ -1058,39 +1060,76 @@ class Nf3 {
     };
 
     connection.onmessage = async message => {
-      const msg = JSON.parse(message.data);
-      const { type, txDataToSign, block, transactions, data } = msg;
+      if (connection._ws && connection._ws.readyState === WebSocket.OPEN) {
+        const msg = JSON.parse(message.data);
+        const { type, txDataToSign, block, transactions, data } = msg;
+
+        logger.debug(`Proposer received websocket message of type ${type}`);
+
+        if (type === 'block') {
+          // First sign transaction, and send it within asynchronous queue. This will
+          // ensure that blockProposed events are emitted in order and with the correct nonce.
+          const tx = await this._signTransaction(
+            txDataToSign,
+            this.stateContractAddress,
+            await this.getBlockStake(), // the block stake could have changed, so we get it from the blockchain
+          );
+          proposerQueue.push(async () => {
+            try {
+              const receipt = await this._sendTransaction(tx);
+              proposeEmitter.emit('receipt', receipt, block, transactions);
+            } catch (err) {
+              logger.error({
+                msg: 'Error while trying to submit a block',
+                err,
+              });
+
+              // block proposed is reverted. Send transactions back to mempool
+              try {
+                await axios.get(`${this.optimistBaseUrl}/block/reset-localblock`);
+              } catch (errorResetLocalBlock) {
+                logger.error({
+                  msg: 'Error while trying to reset local block',
+                  errorResetLocalBlock,
+                });
+              }
+              proposeEmitter.emit('error', err, block, transactions);
+            }
+          });
+        }
 
-      logger.debug(`Proposer received websocket message of type ${type}`);
+        if (type === 'rollback') proposeEmitter.emit('rollback', data);
 
-      if (type === 'block') {
-        // First sign transaction, and send it within asynchronous queue. This will
-        // ensure that blockProposed events are emitted in order and with the correct nonce.
-        const tx = await this._signTransaction(txDataToSign, this.stateContractAddress, 0); // we don't send more stake
-        proposerQueue.push(async () => {
+        // this is used by adversary proposer for submitting bad transaction.
+        if (type === 'submit-transaction') {
```

Comment:
This is not my code.

Reply:
This was added because there is a test that constantly reboots the node. This additional code helps to support this scenario. In any case, tjñjis code is going away